client/packages/host/src/Admin/WebAppLogFileModal.tsx

```diff
@@ -0,0 +1,145 @@
+
+export const LogDisplay = ({
+  fileName,
+  setLogContent,
+}: {
+  fileName: string;
+  setLogContent: (content: string[]) => void;
+}) => {
+  const { mutateAsync, data, isLoading } =
+    useLog.document.logContentsByFileName();
+
+  useEffect(() => {
+    if (!!fileName) {
+      mutateAsync(fileName);
+    }
+  }, [fileName]);
+
+  useEffect(() => {
+    if (!!data?.fileContent) {
+      setLogContent(data.fileContent);
+    }
+  }, [data]);
+
+  if (isLoading) {
+    return <BasicSpinner />;
+  }
+
+  return !!data?.fileContent ? (
+    <Box paddingTop={2} maxHeight={400}>
+      <LogTextDisplay logText={data?.fileContent} />
+    </Box>
+  ) : null;
+};
+
+export const WebAppLogFileModal = ({
+  isOpen,
+  onClose,
+}: {
+  isOpen: boolean;
+  onClose: () => void;
+}) => {
+  const t = useTranslation();
+  const { success, warning } = useNotification();
+  const [logToRender, setLogToRender] = useState('');
+  const [logContent, setLogContent] = useState<string[]>([]);
+  const { Modal } = useDialog({ isOpen });
+  const { saveFile } = useWebClient();
+  const [isSaving, setIsSaving] = useState(false);
+  const noLog = logContent.length === 0;
+
+  const { data, isError, isLoading } = useLog.document.listFileNames();
+
+  const saveLog = async () => {
+    if (noLog) {
+      warning(t('message.nothing-to-save'))();
+    } else if (isSaving) {
+      warning(t('message.already-saving'))();
+    } else {
+      setIsSaving(true);
+      saveFile({
+        content: logContent.toString(),
+        filename: logToRender,
+      });
+      setIsSaving(false);
+    }
+  };
+
+  const copyToClipboard = () => {
+    navigator.clipboard.writeText(logContent.toString()).then(() => {
+      success(t('message.copy-success'))();
+    });
+  };
+
+  if (isError) {
+    return (
+      <Modal
+        title={t('heading.server-log')}
+        okButton={<DialogButton variant="ok" onClick={onClose} />}
+      >
+        <Box sx={{ padding: 2 }} textAlign="center">
+          <Typography>{t('error.unable-to-load-server-log')}</Typography>
+        </Box>
+      </Modal>
+    );
+  }
+
+  return (
+    <Modal
+      title={t('heading.server-log')}
+      okButton={<DialogButton variant="ok" onClick={onClose} />}
+      width={850}
+      height={700}
+      copyButton={
+        <DialogButton
+          variant="copy"
+          onClick={
+            noLog
+              ? () => warning(t('message.nothing-to-copy'))()
+              : copyToClipboard
+          }
+          color="primary"
+        />
+      }
+      saveButton={
+        <DialogButton variant="save" onClick={saveLog} color="primary" />
+      }
+    >
+      {!isLoading ? (
+        <>
+          <DropdownMenu
+            label={logToRender ? logToRender : t('label.server-log')}
+          >
+            {data?.fileNames?.map((fileName, i) => (
+              <DropdownMenuItem
+                key={i}
+                onClick={() => {
+                  setLogToRender(fileName);
```

Comment:
> I suggest adding this - otherwise when you close and reopen the modal, the state remains and you can save, even though the modal is loading a log and there's nothing displayed

Nice! I like it. Though I am confused by this behaviour. Does the dropdown menu item register an onClick action when the modal is re opened?

Reply:
no, it doesn't. I wasn't clear about the steps I was taking soz
- open modal
- select log
- log displays, save is enabled
- close modal
- open modal
- select log
- log is loading.. nothing is showing.. save is enabled! it saves the previous logfile

so the fix works for that use-case. Another option is to reset the state when the modal is closed, but I thought that resetting when you select a new log is better, in that it means when you select a new log, the save is only available after the log is loaded.